### PR TITLE
dist/common/scripts/scylla_util.py: Add Azure support

### PR DESCRIPTION
**Pull request by yaronkaikov**
In scylladb/scylla-machine-image#109 we need to
add support to Scylla-machine-image on Azure.

Let's add to scylla_util.py the relevant configuration to support Azure as well

Need to get #8344 merged before

### CODE REVIEW

File: dist/common/scripts/scylla_util.py [hunk 1]
```diff
@@ -559,13 +792,18 @@ def is_gce():
...
 def get_cloud_instance():
     if is_ec2():
         return aws_instance()
     elif is_gce():
         return gcp_instance()
+    elif is_azure():
+        return azure_instance()
     else:
-        raise Exception("Unknown cloud provider! Only AWS/GCP supported.")
+        raise Exception("Unknown cloud provider! Only AWS/GCP and AZURE supported.")
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:806]: `Unknown cloud provider. Only AWS, GCP, and Azure are supported.`

File: dist/common/scripts/scylla_io_setup [hunk 2]
```diff
@@ -122,6 +122,14 @@ class scylla_cpuinfo:
             return len(self._cpu_data["system"])
 
 def run_iotune():
+            with open("/etc/sysctl.conf", "r+") as file:
+                for line in file:
+                    if 'aio-max-nr' in line:
+                        break
+                    else: # not found, we are at the eof
+                        file.write("fs.aio-max-nr = 30000000\n") # append missing data
+                        os.system('sysctl -p /etc/sysctl.conf')
```
Comment (penberg) [inline dist/common/scripts/scylla_io_setup:131]: This is generic code so let's be extra careful. Why is this change needed? Perhaps make it a separate pull request.

File: dist/common/scripts/scylla_util.py [hunk 3]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+        self.__type = None
+        self.__cpu = None
+        self.__memoryGB = None
+        self.__nvmeDiskCount = None
+        self.__firstNvmeSize = None
+        self.__osDisks = None
+
+    @staticmethod
+    def is_azure_instance():
+        """Check if it's AZURE instance via DNS lookup to metadata server."""
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:139]: Azure

File: dist/common/scripts/scylla_util.py [hunk 4]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+        try:
+            addrlist = socket.getaddrinfo('169.254.169.254', 80)
+        except socket.gaierror:
+            return False
+        for res in addrlist:
+            af, socktype, proto, canonname, sa = res
+            if af == socket.AF_INET:
+                addr, port = sa
+                if addr == "169.254.169.254":
+                    return True
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:150]: Can you point me to the documentation that describes how to detect Azure? IOW, what you used as a source to implement this check.

File: dist/common/scripts/scylla_util.py [hunk 5]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+    def _non_root_nvmes(self):
+        """get list of nvme disks from os, filter away if one of them is root"""
+        nvme_re = re.compile(r"nvme\d+n\d+$")
+        root_dev_candidates = [x for x in psutil.disk_partitions() if x.mountpoint == "/"]
+        if len(root_dev_candidates) != 1:
+            raise Exception("found more than one disk mounted at root ".format(root_dev_candidates))
+
+        root_dev = root_dev_candidates[0].device
+        # if root_dev.startswith("/dev/mapper"):
+        #     raise Exception("mapper used for root, not checking if nvme is used ".format(root_dev))
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:165]: Please remove commented out code.

File: dist/common/scripts/scylla_util.py [hunk 6]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+
+    def __instance_metadata(self, path, recursive=False):
+        return curl(self.META_DATA_BASE_URL + path + "?api-version=2021-01-01&format=text", headers = { "Metadata": "true" })
+
+    def _non_root_nvmes(self):
+        """get list of nvme disks from os, filter away if one of them is root"""
+        nvme_re = re.compile(r"nvme\d+n\d+$")
+        root_dev_candidates = [x for x in psutil.disk_partitions() if x.mountpoint == "/"]
+        if len(root_dev_candidates) != 1:
+            raise Exception("found more than one disk mounted at root ".format(root_dev_candidates))
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:161]: There's no `{}` marker so there's no place where it could print `root_dev_candidates`.

In any case, please use the modern f-string syntax for string formatting:

```raise Exception(f"Found more than one disk mounted as root: {root_dev_candidates}")
```

File: dist/common/scripts/scylla_util.py [hunk 7]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+            return True
+        return False
+
+    def __get_nvme_disks_from_metadata(self):
+        """get list of nvme disks from metadata server"""
+        import json
+        try:
+            disksREST=self.__instance_metadata("disks", True)
+            disksobj=json.loads(disksREST)
+            nvmedisks=list(filter(self.isNVME, disksobj))
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:198]: Missing spaces before and after assignment operator (`=`).

File: dist/common/scripts/scylla_util.py [hunk 8]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+        """check if disk from AZURE metadata is a NVME disk"""
+        if azurediskobj["interface"]=="NVME":
+            return True
+        return False
+
+    def __get_nvme_disks_from_metadata(self):
+        """get list of nvme disks from metadata server"""
+        import json
+        try:
+            disksREST=self.__instance_metadata("disks", True)
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:196]: Call the variable `raw_disks`, maybe?

File: dist/common/scripts/scylla_util.py [hunk 9]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+            disksobj=json.loads(disksREST)
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:197]: Call the variable, `disks` maybe?

File: dist/common/scripts/scylla_util.py [hunk 10]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+            nvmedisks=list(filter(self.isNVME, disksobj))
+        except Exception as e:
+            print ("Problem when parsing disks from metadata:")
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:200]:
```
print(f"Failed to parse disk metadata: {e}")
```

File: dist/common/scripts/scylla_util.py [hunk 11]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+            nvmedisks={}
+        return nvmedisks
+
+    @property
+    def nvmeDiskCount(self):
+        """get # of nvme disks available for scylla raid"""
+        if self.__nvmeDiskCount is None:
+            try:
+                ephemeral_disks = self.getEphemeralOsDisks()
+                count_os_disks=len(ephemeral_disks)
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:211]: Spaces missing before and after assignment operator (`=`).

File: dist/common/scripts/scylla_util.py [hunk 12]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+            except Exception as e:
+                print ("Problem when parsing disks from OS:")
+                print (e)
+                count_os_disks=0
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:215]: Spaces missing before and after assignment operator (`=`).

File: dist/common/scripts/scylla_util.py [hunk 13]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+            nvme_metadata_disks = self.__get_nvme_disks_from_metadata()
+            count_metadata_nvme_disks=len(nvme_metadata_disks)
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:217]: Spaces missing before and after assignment operator (`=`).

File: dist/common/scripts/scylla_util.py [hunk 14]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+    def is_recommended_instance_size(self):
+        """if this instance has at least 2 cpus, it has a recommended size"""
+        if int(self.instance_size()) > 1:
+            return True
+        return False
+
+    @staticmethod
+    def get_file_size_by_seek(filename):
+        "Get the file size by seeking at end"
+        fd= os.open(filename, os.O_RDONLY)
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:283]: Space missing before assignment operator (`=`).

File: dist/common/scripts/scylla_util.py [hunk 15]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+        try:
+            return os.lseek(fd, 0, os.SEEK_END)
+        finally:
+            os.close(fd)
+
+    # note that GCP has 3TB physical devices actually, which they break into smaller 375GB disks and share the same mem with multiple machines
+    # this is a reference value, disk size shouldn't be lower than that
+    AZURE_NVME_DISK_SIZE_2020=375
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:291]: The comment says GCP. Is there such limit for Azure?

File: dist/common/scripts/scylla_util.py [hunk 16]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+            firstDiskSizeGB = firstDiskSize/1024/1024/1024
+            if firstDiskSizeGB >= self.AZURE_NVME_DISK_SIZE_2020:
+                self.__firstNvmeSize = firstDiskSizeGB
+            else:
+                raise Exception("First nvme is smaller than lowest expected size. ".format(firstDisk))
+        return self.__firstNvmeSize
+
+    def is_recommended_instance(self):
+        if not self.is_unsupported_instance_class() and self.is_supported_instance_class() and self.is_recommended_instance_size():
+            # at least 1:2GB cpu:ram ratio , GCP is at 1:4, so this should be fine
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:309]: Lets either (a) figure out what CPU:RAM ratio we want for Azure (you can talk to @avikivity, for example), or (b) just drop the checks for now.

File: dist/common/scripts/scylla_util.py [hunk 17]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+        return False
+
+    def is_supported_instance_class(self):
+        """Returns if this instance type belongs to supported ones for nvmes"""
+        if self.instancetype == self.m1supported:
+            return True
+        if self.instance_class() in ['n1', 'n2', 'n2d' ,'c2']:
+            return True
+        return False
+
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:272]: Redundant newline, please drop.

File: dist/common/scripts/scylla_util.py [hunk 18]
```diff
@@ -117,6 +117,239 @@ def curl(url, headers=None, byte=False, timeout=3, max_retries=5, retry_interval
...
+    def instance_class(self):
+        """Returns the class of the instance we are running in. i.e.: n2"""
+        return self.instancetype.split("_")[0]
+
+    def instance_purpose(self):
+        """Returns the purpose of the instance we are running in. i.e.: standard"""
+        return self.instancetype.split("_")[1]
+
+    m1supported="m1-megamem-96"
```
Comment (penberg) [inline dist/common/scripts/scylla_util.py:255]: These instance types look like copy-paste from GCP code. Please update this to match Azure instance types.

**Review by penberg — CHANGES_REQUESTED** [review]
Let's also have @syuu1228 and maybe @bentsi to review this.

File: dist/common/scripts/scylla_io_setup (hunk 2 above)
Comment (yaronkaikov) [reply dist/common/scripts/scylla_io_setup:131]: I will remove it as i found there is a function `configure_aio_slots` under `scylla_prepare` .

File: dist/common/scripts/scylla_util.py (hunk 3 above)
Comment (yaronkaikov) [reply dist/common/scripts/scylla_util.py]: Fixed

File: dist/common/scripts/scylla_util.py (hunk 5 above)
Comment (yaronkaikov) [reply dist/common/scripts/scylla_util.py]: Removed

File: dist/common/scripts/scylla_util.py (hunk 8 above)
Comment (yaronkaikov) [reply dist/common/scripts/scylla_util.py]: Changed

File: dist/common/scripts/scylla_util.py (hunk 9 above)
Comment (yaronkaikov) [reply dist/common/scripts/scylla_util.py]: changed

File: dist/common/scripts/scylla_util.py (hunk 11 above)
Comment (yaronkaikov) [reply dist/common/scripts/scylla_util.py]: Fixed

File: dist/common/scripts/scylla_util.py (hunk 12 above)
Comment (yaronkaikov) [reply dist/common/scripts/scylla_util.py]: fixed

File: dist/common/scripts/scylla_util.py (hunk 13 above)
Comment (yaronkaikov) [reply dist/common/scripts/scylla_util.py]: fixed

File: dist/common/scripts/scylla_util.py (hunk 17 above)
Comment (yaronkaikov) [reply dist/common/scripts/scylla_util.py:270]: removed

File: dist/common/scripts/scylla_util.py (hunk 14 above)
Comment (yaronkaikov) [reply dist/common/scripts/scylla_util.py]: fixed

File: dist/common/scripts/scylla_util.py (hunk 1 above)
Comment (yaronkaikov) [reply dist/common/scripts/scylla_util.py]: fixed

**Review by syuu1228 — COMMENTED** [review]
No preconfigured iops/bandwidth for scylla_io_setup for Azure?
Seems like we have supported it on AWS and GCP.

**Comment by tarzanek** [comment]
hmm, clash of PRs ?
https://github.com/scylladb/scylla/pull/7807

**Comment by tarzanek** [comment]
anyhow https://github.com/scylladb/scylla/pull/7807 contains disk IO setup @syuu1228
